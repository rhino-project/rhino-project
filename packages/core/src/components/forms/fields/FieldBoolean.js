import PropTypes from 'prop-types';
import { Input } from 'reactstrap';
import { useController } from 'react-hook-form';
import { useFieldInheritedProps } from 'rhino/hooks/form';
import { useGlobalComponent } from 'rhino/hooks/overrides';

export const FieldBooleanBase = (props) => {
  const { extractedProps, inheritedProps } = useFieldInheritedProps(props);
  const { readOnly, path } = props;
  const {
    field: { ref, value, ...fieldProps },
    fieldState: { error }
  } = useController({
    name: path
  });

  // FIXME: Cannot guarantee that id will be unique
  return (
    <Input
      {...extractedProps}
      {...fieldProps}
      innerRef={ref}
      type="checkbox"
      invalid={!!error}
      checked={value}
      disabled={readOnly}
      {...inheritedProps}
    />
  );
};

FieldBooleanBase.propTypes = {
  path: PropTypes.string.isRequired
};

const FieldBoolean = (props) =>
  useGlobalComponent('FieldBoolean', FieldBooleanBase, props);

export default FieldBoolean;
