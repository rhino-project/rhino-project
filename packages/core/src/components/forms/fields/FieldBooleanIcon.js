import PropTypes from 'prop-types';
import { useController } from 'react-hook-form';
import { useFieldInheritedProps } from '../../../hooks/form';
import { Icon } from '../../icons';
import { useGlobalComponent } from '../../../hooks/overrides';

export const FieldBooleanIconBase = ({
  trueIcon = 'check',
  falseIcon = 'x',
  ...props
}) => {
  const { extractedProps, inheritedProps } = useFieldInheritedProps(props);
  const { path } = props;
  const {
    field: { value, onChange, ...fieldProps }
  } = useController({
    name: path
  });

  return (
    <Icon
      {...extractedProps}
      icon={value ? trueIcon : falseIcon}
      onClick={() => onChange(!value)}
      {...fieldProps}
      {...inheritedProps}
    />
  );
};

FieldBooleanIconBase.propTypes = {
  path: PropTypes.string.isRequired
};

export const FieldBooleanIcon = (props) =>
  useGlobalComponent('FieldBooleanIcon', FieldBooleanIconBase, props);
