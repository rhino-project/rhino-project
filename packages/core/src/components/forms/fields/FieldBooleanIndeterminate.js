import PropTypes from 'prop-types';
import { Input } from 'reactstrap';

import { useController } from 'react-hook-form';
import { useCallback, useEffect, useMemo, useRef } from 'react';
import { useFieldInheritedProps } from '../../../hooks/form';
import { useGlobalComponent } from '../../../hooks/overrides';

// FIXME: This is duplicated from in ModelFilterBoolean
// Ensure that if the value is a string coming from the url, it is either 'true' or 'false'
const parseBooleanFilterValue = (value) => {
  if (typeof value === 'boolean') return value;
  if (typeof value !== 'string') return null;

  const normalized = value.trim().replace(/ /g, '').toLowerCase();
  if (normalized === 'true') return true;
  if (normalized === 'false') return false;
  return null;
};

export const FieldBooleanIndeterminateBase = (props) => {
  const { extractedProps, inheritedProps } = useFieldInheritedProps(props);
  const { readOnly, path } = props;
  const {
    field: { ref, value, ...fieldProps },
    fieldState: { error }
  } = useController({
    name: path
  });
  const indeterminateRef = useRef(null);

  const parsedValue = parseBooleanFilterValue(value);
  const indeterminate = useMemo(
    () => parsedValue !== true && parsedValue !== false,
    [parsedValue]
  );

  // Necessary to set indeterminate state
  // https://react-hook-form.com/faqs#Howtosharerefusage
  const innerRef = useCallback(
    (e) => {
      ref(e);
      indeterminateRef.current = e;
    },
    [ref]
  );

  useEffect(() => {
    indeterminateRef.current.indeterminate = indeterminate;
  }, [indeterminateRef, indeterminate]);

  // FIXME: Cannot guarantee that id will be unique
  return (
    <Input
      {...extractedProps}
      {...fieldProps}
      type="checkbox"
      innerRef={innerRef}
      invalid={!!error}
      checked={parsedValue === true}
      disabled={readOnly}
      {...inheritedProps}
    />
  );
};

FieldBooleanIndeterminateBase.propTypes = {
  path: PropTypes.string.isRequired
};

export const FieldBooleanIndeterminate = (props) =>
  useGlobalComponent(
    'FieldBooleanIndeterminate',
    FieldBooleanIndeterminateBase,
    props
  );
