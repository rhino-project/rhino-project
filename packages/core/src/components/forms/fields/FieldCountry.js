import PropTypes from 'prop-types';
import classnames from 'classnames';

import { useController } from 'react-hook-form';
import { useMemo } from 'react';
import { useGlobalComponent } from 'rhino/hooks/overrides';
import { CloseButton } from 'rhino/components/buttons';
import PhoneInput from 'react-phone-input-2';

export const Flag = ({ country }) => {
  const lowerValue = useMemo(() => country?.toLowerCase(), [country]);

  if (!lowerValue) return null;

  return (
    <div className="country-flag-display react-tel-input d-inline-block ms-2">
      <div className={classnames('flag', lowerValue)} />
    </div>
  );
};

Flag.propTypes = {
  country: PropTypes.string.isRequired
};

export const FieldCountryBase = ({ required, ...props }) => {
  const { path } = props;
  const {
    field: { value, onChange, ...fieldProps },
    fieldState: { error }
  } = useController({
    name: path
  });

  const handleCountryChange = (_, country) => onChange(country.countryCode);
  const handleClear = () => onChange('');

  const lowerValue = useMemo(() => value?.toLowerCase(), [value]);
  const upperValue = useMemo(() => value?.toUpperCase(), [value]);

  return (
    <div className={classnames('country-field', { 'is-invalid': error })}>
      <div className="country-label">
        {upperValue}
        {!required && upperValue && <CloseButton onClick={handleClear} />}
      </div>
      <PhoneInput
        {...fieldProps}
        containerClass={classnames('country-input', 'form-control', {
          'is-invalid': error
        })}
        country={lowerValue}
        preferredCountries={['us', 'ca']}
        regions={['north-america', 'europe']}
        onChange={handleCountryChange}
        enableSearch
        isValid
        {...props}
      />
    </div>
  );
};

FieldCountryBase.propTypes = {
  path: PropTypes.string.isRequired
};

const FieldCountry = (props) =>
  useGlobalComponent('FieldCountry', FieldCountryBase, props);

export default FieldCountry;
