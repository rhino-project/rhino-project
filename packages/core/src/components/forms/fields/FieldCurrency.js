import PropTypes from 'prop-types';
import classnames from 'classnames';
import { useFieldError, useFieldInheritedProps } from '../../../hooks/form';
import { useController } from 'react-hook-form';
import { Input, InputGroup } from 'reactstrap';
import { useCallback, useMemo, useRef } from 'react';
import { applyCurrencyMask, applyCurrencyMaskFromInput } from '../../../utils/ui';
import { useGlobalComponent } from '../../../hooks/overrides';

export const FieldCurrencyBaseInput = ({ ...props }) => {
  const { path } = props;
  const { extractedProps, inheritedProps } = useFieldInheritedProps(props);
  const inputRef = useRef(null);
  const {
    field: { onChange, value: fieldValue, ...fieldProps },
    fieldState: { error }
  } = useController({
    name: path
  });

  const handleOnChange = useCallback((event) => {
    const formattedValue = applyCurrencyMaskFromInput(event);
    if (inputRef.current) {
      inputRef.current.value = formattedValue.value;
      inputRef.current.setSelectionRange(
        formattedValue.selectionStart,
        formattedValue.selectionEnd
      );
    }
  }, []);

  const handleOnBlur = useCallback(() => {
    const value = applyCurrencyMask(inputRef.current?.value);
    onChange(value);
    if (inputRef.current) inputRef.current.value = value;
  }, [onChange]);

  const value = useMemo(() => applyCurrencyMask(fieldValue), [fieldValue]);

  return (
    <Input
      {...extractedProps}
      {...fieldProps}
      defaultValue={value}
      autoComplete="off"
      innerRef={inputRef}
      invalid={!!error}
      onChange={handleOnChange}
      onBlur={handleOnBlur}
      {...inheritedProps}
    />
  );
};

export const FieldCurrencyBase = ({ ...props }) => {
  const { path } = props;
  const error = useFieldError(path);

  return (
    <InputGroup
      className={classnames({
        'is-invalid': error
      })}
    >
      <span className="input-group-text">$</span>
      <FieldCurrencyBaseInput {...props} />
    </InputGroup>
  );
};

FieldCurrencyBase.propTypes = {
  path: PropTypes.string.isRequired
};

export const FieldCurrency = (props) =>
  useGlobalComponent('FieldCurrency', FieldCurrencyBase, props);
