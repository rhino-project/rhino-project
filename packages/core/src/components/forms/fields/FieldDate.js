import PropTypes from 'prop-types';
import DatePicker from 'react-datepicker';
import classnames from 'classnames';

import { useController } from 'react-hook-form';
import { useMemo } from 'react';
import { useFieldInheritedProps } from '../../../hooks/form';
import { useGlobalComponent } from '../../../hooks/overrides';

export const FieldDateBase = ({ min, max, ...props }) => {
  const { placeholder, path } = props;
  const {
    field: { value, onChange, ...fieldProps },
    fieldState: { error }
  } = useController({
    name: path
  });

  // The date will be interpreted as UTC unless we break it apart
  const date = useMemo(() => {
    if (!value) return undefined;

    const parts = value.split('-');
    const year = parseInt(parts[0], 10);
    // JavaScript counts months from 0 to 11
    const month = parseInt(parts[1], 10) - 1;
    const day = parseInt(parts[2], 10);

    return new Date(year, month, day);
  }, [value]);

  const inheritedPropsOptions = useMemo(
    () => ({
      className: classnames('d-block', 'form-control', { 'is-invalid': error })
    }),
    [error]
  );
  const wrapperClassName = useMemo(
    () => classnames('d-block', { 'is-invalid': error }),
    [error]
  );
  const { extractedProps, inheritedProps } = useFieldInheritedProps(
    props,
    inheritedPropsOptions
  );

  const handleChange = (date) => onChange(date?.toISOString() || null);

  return (
    <DatePicker
      {...extractedProps}
      {...fieldProps}
      wrapperClassName={wrapperClassName}
      selected={date}
      onChange={handleChange}
      dateFormat="MMMM d, yyyy"
      autoComplete="off"
      minDate={min}
      maxDate={max}
      placeholderText={placeholder}
      {...inheritedProps}
    />
  );
};

FieldDateBase.propTypes = {
  path: PropTypes.string.isRequired
};

const FieldDate = (props) =>
  useGlobalComponent('FieldDate', FieldDateBase, props);

export default FieldDate;
