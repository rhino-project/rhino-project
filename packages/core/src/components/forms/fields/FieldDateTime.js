import PropTypes from 'prop-types';
import DatePicker from 'react-datepicker';
import classnames from 'classnames';

import { useController } from 'react-hook-form';
import { useMemo } from 'react';
import { useFieldInheritedProps } from 'rhino/hooks/form';
import { useGlobalComponent } from 'rhino/hooks/overrides';

export const FieldDateTimeBase = ({ min, max, ...props }) => {
  const { path, placeholder } = props;
  const {
    field: { value, onChange, onBlur },
    fieldState: { error }
  } = useController({
    name: path
  });

  const date = useMemo(() => (value ? new Date(value) : undefined), [value]);

  const inheritedPropsOptions = useMemo(
    () => ({
      className: classnames('d-block', 'form-control', { 'is-invalid': error })
    }),
    [error]
  );
  const wrapperClassName = useMemo(
    () => classnames('d-block', { 'is-invalid': error }),
    [error]
  );
  const { extractedProps, inheritedProps } = useFieldInheritedProps(
    props,
    inheritedPropsOptions
  );

  const handleChange = (date) => {
    // https://github.com/Hacker0x01/react-datepicker/issues/1991
    const dateWithoutMilliseconds = date ? new Date(date) : null;
    dateWithoutMilliseconds?.setUTCMilliseconds(0);

    onChange(dateWithoutMilliseconds?.toISOString() || null);
  };

  return (
    <DatePicker
      {...extractedProps}
      wrapperClassName={wrapperClassName}
      selected={date}
      onChange={handleChange}
      // FIXME: On blur does not always work
      onBlur={onBlur}
      dateFormat="MMMM d, yyyy h:mm aa"
      autoComplete="off"
      showTimeSelect
      minDate={min}
      maxDate={max}
      placeholderText={placeholder}
      {...inheritedProps}
    />
  );
};

FieldDateTimeBase.propTypes = {
  path: PropTypes.string.isRequired
};

const FieldDateTime = (props) =>
  useGlobalComponent('FieldDateTime', FieldDateTimeBase, props);

export default FieldDateTime;
