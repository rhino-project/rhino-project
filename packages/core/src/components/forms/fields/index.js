export * from './FieldBoolean';
export * from './FieldBooleanIcon';
export * from './FieldBooleanIndeterminate';
export * from './FieldCountry';
export * from './FieldCurrency';
export * from './FieldDate';
export * from './FieldDateTime';
export * from './FieldFile';
export * from './FieldFloat';
export * from './FieldHidden';
export * from './FieldInput';
export * from './FieldInputControlled';
export * from './FieldInteger';
export * from './FieldPassword';
export * from './FieldPhone';
export * from './FieldSelect';
export * from './FieldSelectControlled';
export * from './FieldString';
export * from './FieldText';
export * from './FieldTime';
export * from './FieldYear';
