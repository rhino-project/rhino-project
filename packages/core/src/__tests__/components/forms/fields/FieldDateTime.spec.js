import { render } from '@testing-library/react';
import { FormProvider, useForm } from 'react-hook-form';
import { FieldDateTime } from '../../../../components/forms/fields/FieldDateTime';
import { sharedFieldTests } from './sharedFieldTests';
import { createWrapper } from '../../../shared/helpers';

describe('FieldDateTime', () => {
  const FormWrapper = ({ children, ...props }) => {
    const methods = useForm(props);
    return <FormProvider {...methods}>{children}</FormProvider>;
  };

  sharedFieldTests(FieldDateTime);

  it('renders empty with null date', () => {
    const { asFragment } = render(<FieldDateTime path="dummy" />, {
      wrapper: createWrapper(FormWrapper, {
        defaultValues: { dummy: null }
      })
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders empty with empty string date', () => {
    const { asFragment } = render(<FieldDateTime path="dummy" />, {
      wrapper: createWrapper(FormWrapper, {
        defaultValues: { dummy: '' }
      })
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders error', () => {
    const { asFragment } = render(<FieldDateTime path="dummy" />, {
      wrapper: createWrapper(FormWrapper, {
        defaultValues: { dummy: '' },
        errors: { dummy: { message: 'Error' } }
      })
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
