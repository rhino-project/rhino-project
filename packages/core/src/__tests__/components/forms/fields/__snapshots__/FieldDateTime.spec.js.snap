// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FieldDateTime > render with disabled 1`] = `
<DocumentFragment>
  <div
    class="react-datepicker-wrapper d-block"
  >
    <div
      class="react-datepicker__input-container"
    >
      <input
        autocomplete="off"
        class="d-block form-control"
        disabled=""
        id="dummy"
        name="dummy"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`FieldDateTime > render with place holder 1`] = `
<DocumentFragment>
  <div
    class="react-datepicker-wrapper d-block"
  >
    <div
      class="react-datepicker__input-container"
    >
      <input
        autocomplete="off"
        class="d-block form-control"
        id="dummy"
        name="dummy"
        placeholder="placeholder"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`FieldDateTime > renders empty with empty string date 1`] = `
<DocumentFragment>
  <div
    class="react-datepicker-wrapper d-block"
  >
    <div
      class="react-datepicker__input-container"
    >
      <input
        autocomplete="off"
        class="d-block form-control"
        id="dummy"
        name="dummy"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`FieldDateTime > renders empty with null date 1`] = `
<DocumentFragment>
  <div
    class="react-datepicker-wrapper d-block"
  >
    <div
      class="react-datepicker__input-container"
    >
      <input
        autocomplete="off"
        class="d-block form-control"
        id="dummy"
        name="dummy"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`FieldDateTime > renders inside of model context 1`] = `
<DocumentFragment>
  <div
    class="react-datepicker-wrapper d-block"
  >
    <div
      class="react-datepicker__input-container"
    >
      <input
        autocomplete="off"
        class="d-block form-control"
        id="dummy"
        name="dummy"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`FieldDateTime > should render with global override shorthand 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;
